### PR TITLE
Add mock service

### PR DESCRIPTION
**Pull request by jantimar**
Add IAPServiceMock for mock products
Add IAPProduct as a wrapper for SKProduct

### CODE REVIEW

File: Package.resolved [hunk 1]
```diff
@@ -3,11 +3,11 @@
     "pins": [
       {
         "package": "OpenSSL",
-        "repositoryURL": "https://github.com/IBM-Swift/OpenSSL",
+        "repositoryURL": "https://github.com/industrialbinaries/OpenSSL",
```
Comment (VojtaStavik) [inline Package.resolved:15]: 👍

File: README.md [hunk 2]
```diff
@@ -1,2 +1,28 @@
 # youda
-WIP: Library for IAPurchase
+
+[![SPM Compatible](https://img.shields.io/badge/spm-compatible-brightgreen.svg?style=flat)](https://swift.org/package-manager)
+![Platforms: macOS](https://img.shields.io/badge/platforms-macOS-brightgreen.svg?style=flat)
+[![Twitter](https://img.shields.io/badge/twitter-@i_binaries-blue.svg?style=flat)](https://twitter.com/i_binaries)
+
+Library for help you with In-App-Purchase with your project
```
Comment (VojtaStavik) [inline README.md]: `A helper library for testing In-App purchases in your project`

File: Sources/youda/Extensions/SKProduct+Local.swift [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+//
+//  Copyright (c) 2020 Industrial Binaries
+//  MIT license, see LICENSE file for details
+//
+
+import StoreKit
+
+extension SKProduct {
+  /// The cost of the product formatted in the local currency.
+  var localPrice: String? {
```
Comment (VojtaStavik) [inline Sources/youda/Extensions/SKProduct+Local.swift]: I think the formatter should be an optional parameter here to allow you to change the formatting if needed.

File: Sources/youda/Extensions/SKProduct+Local.swift [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    let formatter = NumberFormatter()
+    formatter.numberStyle = .currency
+    formatter.locale = priceLocale
+    return formatter.string(from: price)
+  }
+}
+
+extension SKProductSubscriptionPeriod {
+  /// - returns: The period off the subscription formatted in the local calendar/language.
```
Comment (VojtaStavik) [inline Sources/youda/Extensions/SKProduct+Local.swift]: `off` -> `of`

File: Sources/youda/Extensions/SKProduct+Local.swift [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  var localPeriod: String? {
```
Comment (VojtaStavik) [inline Sources/youda/Extensions/SKProduct+Local.swift]: The date formatter should be an optional input here, too.

File: Sources/youda/IAPService.swift [hunk 6]
```diff
@@ -8,28 +8,33 @@
...
 
-public struct InAppProductId: Hashable {
-  let identifier: String
-}
-
 extension Notification.Name {
-  static var subscriptionChange: Notification.Name {
+  /// Notification trigged when update bought products
+  public static var subscriptionChange: Notification.Name {
     Notification.Name(rawValue: "co.industrial-binaries.youda.subscription-change")
```
Comment (VojtaStavik) [inline Sources/youda/IAPService.swift]: `.init(rawValue` ??

File: Sources/youda/IAPService.swift [hunk 7]
```diff
@@ -8,28 +8,33 @@
...
 
-  /// Available products from apple developer
-  private(set) var availableProducts = [SKProduct]()
-  /// Purchased products
-  private(set) var purchasedProducts = Set<InAppProductId>()
+  /// Available products from iTunesConnect, it is all products which you can buy from your app
+  public private(set) var availableProducts = [IAPProduct]()
+  /// Purchased products from available products
+  public var purchasedProducts: [IAPProduct] {
+    return availableProducts
```
Comment (VojtaStavik) [inline Sources/youda/IAPService.swift]: omit `return`?

File: Sources/youda/IAPService.swift [hunk 8]
```diff
@@ -39,78 +44,81 @@ final class InAppPurchases: NSObject {
...
-    ) { _ in
-      SKPaymentQueue.default().remove(self)
-    }
   }
 
   private func addPurchasedProduct(identifier: String) {
-    purchasedProducts.insert(InAppProductId(identifier: identifier))
-    // inform about subscription change
+    purchasedProductsIdentifiers.insert(identifier)
+    // Send notification to inform about change purchased products
```
Comment (VojtaStavik) [inline Sources/youda/IAPService.swift]: `Send a notification to inform about changed purchased products`

File: Sources/youda/IAPServiceDelegate.swift [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+//  MIT license, see LICENSE file for details
+//
+
+import Foundation
+
+/// IAP Service delegate for IAP updates
+public protocol IAPServiceDelegate: class {
+  /// Call when `IAPService` update `purchasedProducts`, this function call as many times as add new product
+  /// - Parameter purchasedProducts: Set of current `purchasedProducts`
+  func didUpdate(purchasedProducts: [IAPProduct])
```
Comment (VojtaStavik) [inline Sources/youda/IAPServiceDelegate.swift]: Usually, the first parameter in the delegate call is the caller of the method -> `IAPService`

File: Sources/youda/IAPServiceProtocol.swift [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+/// IAP youda environment
+enum IAPEnvironment {
+  /// Default environment mean sandbox when you launch app from Xcode on your iPhone or production in case when you download app from AppStore, when you try app simulator, it will use `mock` environemnts
+  case `default`
+  /// Mock environment for debug app in simulator
+  /// - Parameters:
+  ///   - availableProducts: Array of mock products which will be return when you try bought product, in case it is nil will return default "Test product"
+  ///   - purchasedProducts: Array of purchased products
+  case mock(available: [IAPProduct]?, purchased: [IAPProduct]?)
```
Comment (VojtaStavik) [inline Sources/youda/IAPServiceProtocol.swift]: Optional arrays?? 😱

File: Sources/youda/IAPServiceProtocol.swift [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+extension IAPServiceProtocol {
+  /// Initialize new instance of `IAPServiceProtocol` base on current environments
+  /// - Parameters:
+  ///   - products: Products for request from apple developer acount
+  ///   - deviceID: Device ID for validate receipt hash
+  ///   - environment: Environment of instance
+  static func configureService(products: Set<String>, deviceID: UUID?, environment: IAPEnvironment = .default) -> IAPServiceProtocol {
```
Comment (VojtaStavik) [inline Sources/youda/IAPServiceProtocol.swift:42]: Why is deviceID optional?

**Review by VojtaStavik — APPROVED** [review]

File: Sources/youda/IAPServiceProtocol.swift (hunk 11 above)
Comment (jantimar) [reply Sources/youda/IAPServiceProtocol.swift:42]: bc `identifierForVendor` is optional https://developer.apple.com/documentation/uikit/uidevice/1620059-identifierforvendor

**Pull request merged**